File: sdk/rs/src/limits.rs
//! Put validation limits and shared transport budgets.

use buffa::{encoding::varint_len, types::bytes_encoded_len};

pub use crate::keys::MAX_KEY_LEN;

pub const MAX_PUT_ENTRIES: usize = 2_000_000;
pub const MAX_REQUEST_MESSAGE_BYTES: usize = 256 * 1024 * 1024;
pub const MAX_VALUE_LEN: usize = 32 * 1024 * 1024;

// A stored batch adds sequence metadata to a full-size request. JSON reads also add
// base64 expansion, so responses get double the request budget.
pub const MAX_RESPONSE_MESSAGE_BYTES: usize = 2 * MAX_REQUEST_MESSAGE_BYTES;
pub const MAX_RESPONSE_ELEMENT_MEMORY_BYTES: usize = 256 * 1024 * 1024;

pub const INGEST_ERROR_DOMAIN: &str = "log.ingest";
pub const PUT_TOO_LARGE_REASON: &str = "PUT_TOO_LARGE";

/// Encoded contribution of one physical key/value pair to a protobuf `PutRequest`.
/// Includes the repeated entry framing and omits empty byte fields.
pub fn put_entry_encoded_len(key: &[u8], value: &[u8]) -> usize {
    // Proto3 omits empty byte fields. All three field tags occupy one byte.
    let key_len = if key.is_empty() {
        0
    } else {
        1 + bytes_encoded_len(key)
    };
    let value_len = if value.is_empty() {
        0
    } else {
        1 + bytes_encoded_len(value)
    };
    let entry_len = key_len + value_len;
    1 + varint_len(entry_len as u64) + entry_len
}

/// Exact encoded size of a protobuf `PutRequest` carrying `kvs`.
pub fn put_encoded_len<'a>(kvs: impl IntoIterator<Item = (&'a [u8], &'a [u8])>) -> usize {
    kvs.into_iter()
        .map(|(key, value)| put_entry_encoded_len(key, value))
        .sum()
}

/// The entry count and limit reported by an ingest size rejection.
#[derive(Debug, Clone, Copy, Eq, PartialEq, thiserror::Error)]
#[error("put has {entries} entries, exceeding the limit of {max_entries}")]
pub struct PutTooLarge {
    pub entries: usize,
    pub max_entries: usize,
}

File: sdk/ts/src/limits.ts
export const MAX_PUT_ENTRIES = 2_000_000;
export const MAX_REQUEST_MESSAGE_BYTES = 256 * 1024 * 1024;
export const MAX_VALUE_LEN = 32 * 1024 * 1024;
export const MAX_KEY_LEN = 254;

export type PutEncoding = 'binary' | 'json';

export interface PutLimits {
    maxEntries?: number;
    maxEncodedBytes?: number;
    maxValueLen?: number;
}

export interface PutBatchOptions extends PutLimits {
    encoding?: PutEncoding;
}

type PutEntry = { key: Uint8Array; value: Uint8Array };

export function normalizePutOptions(options: PutBatchOptions = {}): Required<PutBatchOptions> {
    const limits = {
        maxEntries: options.maxEntries ?? MAX_PUT_ENTRIES,
        maxEncodedBytes: options.maxEncodedBytes ?? MAX_REQUEST_MESSAGE_BYTES,
        maxValueLen: options.maxValueLen ?? MAX_VALUE_LEN,
    };
    for (const [name, value] of Object.entries(limits)) {
        if (!Number.isSafeInteger(value) || value < (name === 'maxValueLen' ? 0 : 1)) {
            throw new RangeError(`${name} must be a ${name === 'maxValueLen' ? 'nonnegative' : 'positive'} safe integer`);
        }
    }
    return { ...limits, encoding: options.encoding ?? 'json' };
}

function varintLen(value: number): number {
    let length = 1;
    while (value >= 128) {
        value = Math.floor(value / 128);
        length++;
    }
    return length;
}

export function putEntryEncodedLen(entry: PutEntry, encoding: PutEncoding): number {
    const key = entry.key.byteLength;
    const value = entry.value.byteLength;
    if (encoding === 'binary') {
        const length = (key === 0 ? 0 : 1 + varintLen(key) + key)
            + (value === 0 ? 0 : 1 + varintLen(value) + value);
        return 1 + varintLen(length) + length;
    }

    // ProtoJSON omits empty byte fields and uses padded base64 for the others.
    return 2 + (key === 0 ? 0 : 8 + 4 * Math.ceil(key / 3))
        + (value === 0 ? 0 : 10 + 4 * Math.ceil(value / 3))
        + (key !== 0 && value !== 0 ? 1 : 0);
}

export function putMessageEncodedLen(entryBytes: number, count: number, encoding: PutEncoding): number {
    if (encoding === 'binary') return entryBytes;
    return count === 0 ? 2 : 9 + count + entryBytes;
}

export function putEncodedLen(entries: readonly PutEntry[], encoding: PutEncoding = 'json'): number {
    let entryBytes = 0;
    for (const entry of entries) entryBytes += putEntryEncodedLen(entry, encoding);
    return putMessageEncodedLen(entryBytes, entries.length, encoding);
}

export function validatePutEntry(entry: PutEntry, index: number, maxValueLen: number): void {
    if (entry.key.byteLength > MAX_KEY_LEN) {
        throw new RangeError(`Put entry ${index} key length ${entry.key.byteLength} exceeds ${MAX_KEY_LEN}`);
    }
    if (entry.value.byteLength > maxValueLen) {
        throw new RangeError(`Put entry ${index} value length ${entry.value.byteLength} exceeds ${maxValueLen}`);
    }
}

export function validatePut(entries: readonly PutEntry[], options: PutBatchOptions = {}): void {
    const limits = normalizePutOptions(options);
    if (entries.length === 0 || entries.length > limits.maxEntries) {
        throw new RangeError(`Put requires between 1 and ${limits.maxEntries} entries, got ${entries.length}`);
    }
    let entryBytes = 0;
    for (const [index, entry] of entries.entries()) {
        validatePutEntry(entry, index, limits.maxValueLen);
        entryBytes += putEntryEncodedLen(entry, limits.encoding);
    }
    const length = putMessageEncodedLen(entryBytes, entries.length, limits.encoding);
    if (length > limits.maxEncodedBytes) {
        throw new RangeError(`Put encoded size ${length} exceeds ${limits.maxEncodedBytes}`);
    }
}
